src/cachemachine/rubinimageinfo.py

```diff
@@ -0,0 +1,346 @@
+from typing import Optional
+
+import semver
+from docker_image import reference
+
+from .types import IncomparableImagesError
+
+DEFAULT_TAG = "latest"  # Docker convention
+DEFAULT_RECOMMENDED_TAG = "recommended"
+
+
+class RubinImageType(Enum):
+    """Class to capture the different kinds of image types possible in a
+    Rubin Lab Image container, given our standard tag set."""
+
+    RECOMMENDED = auto()
+    LATEST = auto()
+    LATEST_SOMETYPE = auto()
+    RELEASE = auto()
+    WEEKLY = auto()
+    DAILY = auto()
+    RELEASE_CANDIDATE = auto()
+    EXPERIMENTAL = auto()
+    MALFORMED_TAG = auto()
+
+
+# Helper functions that don't rely on any part of the class object and thus
+# are not methods.
+
+
+def _check_tag_is_lowercase(tag: str) -> bool:
+    if tag != tag.lower():
+        return False
+    return True
+
+
+def _type_from_tag(tag: str, recommended_tag: str) -> RubinImageType:
+    """This constructs the image type from the tag.  It is inherently
+    Rubin tag-specific."""
+    if not _check_tag_is_lowercase(tag):
+        return RubinImageType.MALFORMED_TAG
+    if tag == recommended_tag:
+        return RubinImageType.RECOMMENDED
+    if tag == DEFAULT_TAG:
+        return RubinImageType.LATEST
+    if tag.startswith("latest_"):
+        # It's latest *something*
+        ttag = tag[7:]
+        if ttag in ("weekly", "daily", "release"):
+            return RubinImageType.LATEST_SOMETYPE
+        return RubinImageType.MALFORMED_TAG
+    if tag.startswith("w_"):
+        return RubinImageType.WEEKLY
+    if tag.startswith("d_"):
+        return RubinImageType.DAILY
+    if tag.startswith("exp_"):
+        return RubinImageType.EXPERIMENTAL
+    if tag.startswith("r"):
+        # If there are any non-numeric characters in the tag after the inital
+        # 'r', it's a release candidate; if not, we claim it's a real release.
+        # Thus, r22_0_0_rc1 is a release candidate, but r21_0_1 is a release.
+        tagtext = tag[1:].replace("_", "")
+        if tagtext.isdigit():
+            return RubinImageType.RELEASE
+        return RubinImageType.RELEASE_CANDIDATE
+    return RubinImageType.MALFORMED_TAG
+
+
+def _semver_from_tag(tag_type: RubinImageType, tag: str) -> semver.VersionInfo:
+    if (
+        tag_type == RubinImageType.EXPERIMENTAL
+        or tag_type == RubinImageType.RECOMMENDED
+        or tag_type == RubinImageType.LATEST
+        or tag_type == RubinImageType.LATEST_SOMETYPE
+        or tag_type == RubinImageType.MALFORMED_TAG
+    ):
+        # The "latest" tags are by their nature singletons.
+        # Experimentals and malformed tags do not have meaningful
+        # semantic versioning and are compared purely lexigraphically by tag.
+        return semver.VersionInfo(0, 0, 1)
+    tag = _canonicalize_tag(tag_type, tag)
+    tagparts = tag.split("_")
+    major = int(tagparts[1])
+    minor = int(tagparts[2])
+    patch = 0
+    bldidx = 3  # Index of tag part where build version starts
```

Comment:
uh, what?

Reply:
This is the problem with having both weekly and daily tags.  In both weeklies and dailies, the major version is the year, the minor version is the month-or-week, and then there's trouble.

That's because for weeklies, the patch version is 0, and for dailies, it's the day.  But _then_ there can be additional components, which get semverred into the build string, such as the cycle tag or additional descriptors (eventually including the build timestamp).  The problem is that the number of _-separated components before you get there varies depending on the build type, and that's what I'm solving here.  I have to know the type of image before I know which field starts the stuff that goes into the build version.